cmd/prune-junit-xml/prunexml.go

```diff
@@ -82,6 +88,25 @@ func pruneXML(suites *junitxml.JUnitTestSuites, maxBytes int) {
 	}
 }
 
+// This function condenses the junit xml to have testcase element of only top level tests.
+// Tescases of subtests are included if they fail.
+func prunePASSEDSubtests(suites *junitxml.JUnitTestSuites) {
```

Comment:
do we not show a top-level name for failed tests?

we can maybe leave this for later, but I think ideally for testgrid you want to have the same set of top-level identifiers across runs and if we need more granular results we make them nested under the top level names cc @michelle192837

Reply:
Consistency on row names is usually better, yeah. If only failing results are reported, those failing results will create a row (which is the expensive part for storage/display), and weird things might happen with alerting and the like (because something that runs intermittently and fails with no examples of it passing looks like a consistent failure). I think pruning all subtests might be better in that case.

As an additional possibility to trimming all subtests:

If you want to include some more info, I think it'd be possible to report a flaky result...I believe you'd want to report the pass/fail status for each subtest as a common higher-level name, and TestGrid will merge them when processing into one cell. (This is worth testing out somewhere first though).

e.g. iirc imagine you have
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test1"` (pass)
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test2"` (fail)
- `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test3"` (pass)

Report these as `testcase classname="k8s.io/kubernetes/test/my-stuff" name="Test"` instead 3 times, 2 passing and 1 failing, and TestGrid should merge it into a single cell with a purple 'flaky' status and "2/3 runs passed".

Reply:
I'm not sure whether I like reporting "k8s.io/kubernetes/test/my-stuff" as flaky. That is potentially misleading when the "Test2" fails reliably, i.e. there is no flake.

Reply:
> do we not show a top-level name for failed tests?

If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

Reply:
> If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

Right, but currently I only see sub-test entries and no top-level entry https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672

Reply:
> > do we not show a top-level name for failed tests?
> 
> If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.

@pohly Yes, you are right! The parent tests are reporting failure when a sub-test failed. Hence all the subtests can be pruned for test grid to have the same set of rows for all runs as @BenTheElder and @michelle192837 mentioned.
I shall submit a patch to remove the condition to include failed sub-tests

@BenTheElder Below image depicts how top-level test has failed status when sub test fails:
![image](https://user-images.githubusercontent.com/63038004/229990975-db56ac3b-aef1-4e12-b5db-04ceacff74af.png)

Reply:
> > If a sub-test failed, all parent tests are also reported as failed, i.e. won't get pruned.
> 
> Right, but currently I only see sub-test entries and no top-level entry https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672
> 
> Edit: I see we're still doing package + test name, just not golang subtests.
> 
> I think we should be considering package name as the top level and nesting under that, there's already a ridiculous amount of packages and those are more stable top level identifiers

@BenTheElder In the link that you mentioned, https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-1-27/1643306358801436672

I see both top-level and sub-test entries.
![image](https://user-images.githubusercontent.com/63038004/229992226-4461859b-0b82-4488-abc2-14152c7d0b87.png)

Pruning these subtest rows and having only top-level tests would condense the no of rows from ~39k to ~9k.

If a Top-level test indicates failure, we may get further info on what subtest failed by looking at the Build Log, which is generated from job's stdout file and hence has all the details.

Reply:
> I shall submit a patch to remove the condition to include failed sub-tests

Actually, I prefer the current approach of showing exactly which sub-tests have failed also in testgrid.

If I understand @michelle192837 correctly, it will be a bit odd to see such a flaky sub-test in columns where it failed and not in those were it succeeded, but IMHO that still gives a better impression of what's going on than hiding sub-tests entirely.

Reply:
@pohly I have reverted the change. Can you please take a look?

Reply:
Yeah to be clear, it's odd and may mess with any alerting settings you have, but as long as that's understood you're free to do so. (I just want to be very clear because this kind of thing, only showing failures and not passes, has led to a _lot_ of confusion before. It may be worth adding a note about this behavior in the description of the tab, e.g. `description` in https://github.com/GoogleCloudPlatform/testgrid/blob/master/config.md#prow-job-configuration or https://github.com/GoogleCloudPlatform/testgrid/blob/master/config.md#tab-descriptions).